File: src/15-3sum/solution.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
  nums.sort!
  res = []
  nums.each_with_index do |num, i|
    break if num > 0
    next if i > 0 && num == nums[i - 1]

    l, r = i + 1, nums.length - 1
    while l < r
      sum = num + nums[l] + nums[r]
      if sum.zero?
        res << [num, nums[l], nums[r]]
        l += 1
        r -= 1

        l += 1 while l < r && nums[l] == nums[l - 1]
        r -= 1 while l < r && nums[r] == nums[r + 1]
      elsif sum > 0
        r -= 1
      else
        l += 1
      end
    end
  end

  res
end

File: src/15-3sum/solution.ts
function threeSum(nums: number[]): number[][] {
  let lo: number, hi: number, sum: number;
  let n: number = nums.length - 1;
  let res: number[][] = [];

  nums.sort((x, y) => x - y);

  for (let i = 0; i < n - 1; i++) {
    while (nums[i] === nums[i - 1]) i++;

    (lo = i + 1), (hi = n);
    while (lo < hi) {
      sum = nums[i] + nums[lo] + nums[hi];
      if (sum === 0) {
        res.push([nums[i], nums[lo], nums[hi]]);
        lo++;
        hi--;

        while (nums[lo] == nums[lo - 1]) lo++;
        while (nums[hi] == nums[hi + 1]) hi--;
      } else if (sum > 0) {
        hi--;
      } else {
        lo++;
      }
    }
  }

  return res;
}
